controllers/preflightvalidation_controller.go

```diff
@@ -0,0 +1,239 @@
+				modulesToCheck = append(modulesToCheck, module)
+			}
+		} else {
+			log.Info(utils.WarnString("missing status of Module CR in preflight"), "module", module.Name)
+		}
+	}
+	return modulesToCheck, nil
+}
+
+func (r *PreflightValidationReconciler) getModulesValidationStatusesMap(pv *kmmv1beta1.PreflightValidation) map[string]kmmv1beta1.CRStatus {
+	statusMap := make(map[string]kmmv1beta1.CRStatus, len(pv.Status.CRStatuses))
+	for _, status := range pv.Status.CRStatuses {
+		statusMap[status.Name] = status
+	}
+
+	return statusMap
+}
+
+func (r *PreflightValidationReconciler) updatePreflightStatus(ctx context.Context, pv *kmmv1beta1.PreflightValidation, moduleName, message string, verified bool) {
+	log := ctrl.LoggerFrom(ctx)
+	verificationStatus := kmmv1beta1.VerificationFalse
+	if verified {
+		verificationStatus = kmmv1beta1.VerificationTrue
+	}
+
+	moduleStatus := r.getModuleStatus(pv, moduleName)
+	if moduleStatus == nil {
+		log.Info(utils.WarnString("failed to find the status of Module CR in preflight"), "module", moduleName)
+		return
+	}
+	err := r.statusUpdater.PreflightSetVerificationStatus(ctx, pv, moduleStatus, verificationStatus, message)
+	if err != nil {
+		log.Info(utils.WarnString("failed to update the status of Module CR in preflight"), "module", moduleName, "error", err)
+	}
+}
+
+func (r *PreflightValidationReconciler) presetNewModulesStatuses(ctx context.Context, pv *kmmv1beta1.PreflightValidation, modules []kmmv1beta1.Module) error {
+	for _, module := range modules {
+		if module.GetDeletionTimestamp() != nil {
+			continue
+		}
+		status := r.getModuleStatus(pv, module.Name)
+		if status != nil {
+			continue
+		}
+		pv.Status.CRStatuses = append(pv.Status.CRStatuses, kmmv1beta1.CRStatus{Name: module.Name})
+		status = &pv.Status.CRStatuses[len(pv.Status.CRStatuses)-1]
+		err := r.statusUpdater.PreflightPresetVerificationStatus(ctx, pv, status)
+		if err != nil {
+			return fmt.Errorf("failed to preset Module %s: %w", module.Name, err)
+		}
+	}
+	return nil
+}
+
+func (r *PreflightValidationReconciler) checkPreflightCompletion(ctx context.Context, name, namespace string) (bool, error) {
+	pv := kmmv1beta1.PreflightValidation{}
+	err := r.client.Get(ctx, types.NamespacedName{Namespace: namespace, Name: name}, &pv)
+	if err != nil {
+		return false, fmt.Errorf("failed to get preflight validation object in checkPreflightCompletion: %w", err)
+	}
+
+	for _, crStatus := range pv.Status.CRStatuses {
+		if crStatus.VerificationStatus != kmmv1beta1.VerificationTrue {
+			ctrl.LoggerFrom(ctx).Info("at least one Module is not verified yet", "module", crStatus.Name, "status", crStatus.VerificationStatus)
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+func (r *PreflightValidationReconciler) getModuleStatus(pv *kmmv1beta1.PreflightValidation, moduleName string) *kmmv1beta1.CRStatus {
+	for i := range pv.Status.CRStatuses {
+		if pv.Status.CRStatuses[i].Name == moduleName {
+			return &pv.Status.CRStatuses[i]
+		}
+	}
```

Comment:
How about using a a map instead of a slice as @qbarrand suggested?

Reply:
fixed